src/parsers/dict_to_struct.jl

```diff
@@ -57,57 +57,64 @@ function ps_dict2ps_struct(data::Dict{String,Any})
         @warn "key 'services' not found in PowerSystems dictionary, this will result in an empty services array"
     end
     if haskey(data,"forecasts")
-        devices = collect(vcat(buses,generators,storage,branches,loads,loadZones,shunts,branches,services))
-        forecasts = PowerSystems.forecasts_dict_parser(data["forecasts"],devices)
+        devices = vcat(buses,generators,storage,branches,loads,loadZones,shunts,branches,services)
+        forecasts = make_forecast_array(devices, data["forecasts"]["forecasts"])
     else
-        forecasts = Dict{Symbol, Vector{ <: Forecast}}()
+        forecasts = Vector{Forecast}()
     end
 
     return sort!(buses, by = x -> x.number), generators, storage,  sort!(branches, by = x -> x.connectionpoints.from.number), loads, loadZones, shunts, forecasts, services
 
 end
 
 """
-Finds the values matching an identifier within a dictionary and returns a dict with keyed by
-objects with values as arrays of keys to navigate to the object within the input dict
+Recurse through a nested dictionary looking for dicts with the key key_of_interest.
+Return a dictionary where keys are the value of key_of_interest and values are references
+to those dicts.
+
+Primary use is to build a mapping of component name to component within a ps_dict.
+
 """
-function _retrieve(dict::T, key_of_interest::Union{DataType,Union,String,Symbol}, output::Dict, path::Array) where T<:AbstractDict
-    iter_result = Base.iterate(dict)
-    last_element = length(path)
-    while iter_result !== nothing
-        ((key,value), state) = iter_result
-        if isa(key_of_interest,Union{DataType,Union})
-            if typeof(value) <: key_of_interest
-                output[key] = !haskey(output,key) ? path[1:end] : push!(output[key],path[1:end])
-            end
-        elseif key == key_of_interest
-            output[value] = !haskey(output,value) ? path[1:end] : push!(output[value],path[1:end])
-        end
+function retrieve(dict::AbstractDict, key_of_interest)
+    output = Dict()
+    path = Vector()
+    _retrieve(dict, key_of_interest, output)
+    return output
+end
+
+function _retrieve(dict::AbstractDict, key_of_interest, output::Dict)
+    for (key, value) in dict
         if value isa AbstractDict
-            push!(path,key)
-            _retrieve(value, key_of_interest, output, path)
-            path = path[1:last_element]
+            if haskey(value, key_of_interest)
+                output[key] = value
+            else
+                # Recurse.
+                _retrieve(value, key_of_interest, output)
+            end
         end
-        iter_result = Base.iterate(dict, state)
     end
-    return output
 end
 
+# TODO DT: alternate implementation
```

Comment:
I first simplified the existing function to this implementation.  Then I decided what we actually want is a reference to the actual object, so went with the implementation above.  If we actually do want to use the array of keys, let me know.  But I find this simpler.

Something else to think about:  we could change ps_dict along the same lines as the forecasts in this PR.  We could store arrays of component dictionaries instead of nested dictionaries keyed by attributes.  Aside from code complexity, the main problem with the nested dictionaries is that it is hard to add layers.

Reply:
1. we added a yaml file that is supposed to create a mapping between generators to parse the CDM https://github.com/NREL/PowerSystems.jl/pull/164 maybe it is reasonable to do something like this or extend the yaml to account for the forecast? 

2. I am not sure about the rules to modify the ps_dict since we got those from power models. But I certainly support the idea of simpler intermediate dicts.